convert2rhel/actions/__init__.py

```diff
@@ -225,68 +282,273 @@ def get_actions(actions_path, prefix):
+
+        :rtype: FinishedActions
+        .. important:: Success is currently defined as an action whose status after
+            running is WARNING or better (WARNING or SUCCESS) and
+            failure as worse than WARNING (OVERRIDABLE, ERROR, FATAL)
+        """
+        logger.task("Prepare: %s" % self.task_header)
+
+        if self._has_run:
+            raise ActionError("Stage %s has already run." % self.stage_name)
+        self._has_run = True
+
+        # Make a mutable copy of these parameters so we don't overwrite the caller's data.
+        # If they weren't passed in, default to an empty list.
+        successes = [] if successes is None else list(successes)
+        failures = [] if failures is None else list(failures)
+        skips = [] if skips is None else list(skips)
+
+        # When testing for failed dependencies, we need the Action ids of faiures and skips so
+        # record those separately
+        failed_action_ids = set()
+
+        for action_class in resolve_action_order(
+            self.actions, previously_resolved_actions=successes + failures + skips
+        ):
+            # Decide if we need to skip because deps have failed
+            failed_deps = [d for d in action_class.dependencies if d in failed_action_ids]
+
+            action = action_class()
+
+            if failed_deps:
+                to_be = "was"
+                if len(failed_deps) > 1:
+                    to_be = "were"
+                message = "Skipped because %s %s not successful" % (format_sequence_as_message(failed_deps), to_be)
+
+                action.set_result(status="SKIP", error_id="SKIP", message=message)
+                skips.append(action)
+                failed_action_ids.add(action.id)
+                logger.error("Skipped %s. %s" % (action.id, message))
+                continue
+
+            # Run the Action
+            try:
+                action.run()
+            except (Exception, SystemExit) as e:
+                # Uncaught exceptions are handled by constructing a generic
+                # failure message here that should be reported
+                message = (
+                    "Unhandled exception was caught: %s\n"
+                    "Please file a bug at https://issues.redhat.com/ to have this"
+                    " fixed or a specific error message added.\n"
+                    "Traceback: %s" % (e, traceback.format_exc())
+                )
+                action.set_result(status="ERROR", error_id="UNEXPECTED_ERROR", message=message)
+
+            # Categorize the results
+            if action.status <= STATUS_CODE["WARNING"]:
+                logger.info("%s has succeeded" % action.id)
+                successes.append(action)
+
+            if action.status > STATUS_CODE["WARNING"]:
+                ### FIXME: Probably some code in reports that can do this better
+                logger.error("%s has failed (%s): %s" % (action.id, action.status, action.message))
```

Comment:
This is actually the part that Michal was wanting to output success and failure as each Action ran.  I am going to see if there's something from report that can do the formatting/some of rhe formatting for this, though.

Reply:
Oh, right! That's fine then

Reply:
Okay, this piece of code is now upated.  I had to move the `format_report_message()` function from `actions/report.py` to `actions/__init__.py` to avoid a circular import and I moved a little bit of the logic from `sumary()` into `format_report_message()` to make it easier for the two pieces of code to use it together.